src/Sarif/PeekableStream.cs

```diff
@@ -0,0 +1,129 @@
+            this.cursor = 0;
+        }
+
+        public override bool CanRead => true;
+
+        // Rewind() is much weaker than Seek.  We should not advertise a capability we do not have.
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length => underlyingStream.Length;
+
+        public override long Position 
+        {
+            get => Math.Min(underlyingStream.Position, cursor);
+            set => throw new System.NotImplementedException();
+        }
+
+        public override void Flush()
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (this.rewindBuffer == null)
+            {
+                // Common case: we're past the "peekability window" and are now just a thin wrapper around the underlying Stream.
+                return underlyingStream.Read(buffer, offset, count);
+            }
+            else if (this.cursor == this.underlyingStream.Position)
+            {
+                // Buffer cursor is "caught up" to stream position, so we are not in a "rewound" state.
+                if (this.cursor == this.rewindBuffer.Length)
+                {
+                    // A previous read has brought us to the brink of exceeding the "peekability window", and we are
+                    // now reading more.  So we abandon the buffer, and serve the read as a thin wrapper.
+                    this.rewindBuffer = null;
+                    this.cursor = int.MaxValue;
+                    return underlyingStream.Read(buffer, offset, count);
+                }
+                else
+                {
+                    // Read into our rewind buffer for "backup" and then copy into caller buffer.
+                    // Buffer position is behind Stream position.  Service read out of the buffer.
+                    int aspirationalReadCount = Math.Min(count, (int)(this.rewindBuffer.Length - cursor));
+                    int actualReadCount = underlyingStream.Read(this.rewindBuffer, cursor, aspirationalReadCount);
+                    Array.Copy(this.rewindBuffer, cursor, buffer, offset, actualReadCount);
+
+                    cursor += actualReadCount;
+
+                    return actualReadCount;
+                }
+            }
+            else if (this.cursor <= this.underlyingStream.Position)
+            {
+                // We are rewound.  Service the read out of the rewind buffer.
+                int safeReadCount = Math.Min(count, this.rewindBuffer.Length - cursor);
+                safeReadCount = Math.Min(safeReadCount, (int)(this.underlyingStream.Position - cursor));
+                Array.Copy(this.rewindBuffer, cursor, buffer, offset, safeReadCount);
+
+                cursor += safeReadCount;
+                return safeReadCount;
+            }
+            else
```

Comment:
>else [](http://example.com/codeflow?start=12&length=4)

NIT, I would tend to elise this `else` statement, just leaving the throw to close the function. IMO, you don't need an // unreachable comment when you handle it this way. The last statement is a throw, pretty clear.